### PR TITLE
[ADD][8.0] Website portal purchase

### PR DESCRIPTION
**Pull request by sergio-teruel**
This module adds the partner's purchase documents in the frontend portal. Suppliers will be able to see the list and state of their request for quotations, purchase orders and supplier invoices.

### CODE REVIEW

**Comment by yajo** [comment]
I suppose this will fail until merging https://github.com/OCA/server-tools/pull/277, right?

**Comment by yajo** [comment]
Please rebase or merge.

**Comment by sergio-teruel** [comment]
Rebase done after https://github.com/OCA/server-tools/pull/277 merged but dependency error continue

**Comment by rafaelbn** [comment]
cc @pedrobaeza

**Comment by yajo** [comment]
If I'm not wrong, you have to add to .travis.yml [something like this](https://github.com/OCA/crm/blob/9.0/.travis.yml#L26) but pointing to the server-tools repo.

**Comment by rafaelbn** [comment]
Rebuilding for testing more than 2 hours

```
Date    Level   Type    Message
2015-11-12 02:53:42     INFO    runbot  odoo.runbot:0 init Init build environment
2015-11-12 02:53:52     INFO    runbot  odoo.runbot:0 test_base Start pre-build commands
2015-11-12 03:06:52     INFO    runbot  odoo.runbot:0 test_base Start test base module
2015-11-12 03:23:32     INFO    runbot  odoo.runbot:0 test_all Start test all modules 
```

cc @pedrobaeza

File: .travis.yml [hunk 1]
```diff
@@ -37,6 +37,7 @@ before_install:
 
 install:
   - git clone https://github.com/OCA/maintainer-quality-tools.git ${HOME}/maintainer-quality-tools
+  - git clone https://github.com/OCA/server-tools.git -b ${VERSION} ${HOME}/server-tools
```
Comment (pedrobaeza) [inline .travis.yml]: Don't add this here

File: website_portal_purchase/README.rst [hunk 2]
```diff
@@ -0,0 +1,64 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+
+============================
+Website Portal for Purchases
+============================
+
+This module add the user's purchases documents in the frontend portal.
```
Comment (pedrobaeza) [inline website_portal_purchase/README.rst]: s/add/adds

Comment (pedrobaeza) [inline website_portal_purchase/README.rst]: s/user's/partner's

Comment (pedrobaeza) [inline website_portal_purchase/README.rst]: s/purchases/purchase

File: website_portal_purchase/README.rst [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+Suppliers will be able to see the list and state of their quotations, purchases
+and invoices.
```
Comment (pedrobaeza) [inline website_portal_purchase/README.rst]: s/invoice/supplier invoices

File: website_portal_purchase/README.rst [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+Usage
+=====
+
+To use this module, you need to:
+
+1. You must select an account type in signup form.
```
Comment (pedrobaeza) [inline website_portal_purchase/README.rst]: Remove "You must"

Comment (pedrobaeza) [inline website_portal_purchase/README.rst]: Instructions seems incomplete

File: website_portal_purchase/README.rst [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/186/8.0
+
+For further information, please visit:
```
Comment (pedrobaeza) [inline website_portal_purchase/README.rst]: Remove this useless sentence

File: website_portal_purchase/__openerp__.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+# (c) 2015 Antiun Ingeniería S.L. - Sergio Teruel
+# (c) 2015 Antiun Ingeniería S.L. - Carlos Dauden
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+{
+    'name': 'Website Portal for Purchases',
+    'category': 'Website',
+    'summary': (
+        'Add your purchase document in the frontend portal (purchases order'
```
Comment (pedrobaeza) [inline website_portal_purchase/__openerp__.py]: s/purchases/purchase

File: website_portal_purchase/__openerp__.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        ', quotations, invoices)'
```
Comment (pedrobaeza) [inline website_portal_purchase/__openerp__.py]: s/quotations/request for quotations

Comment (pedrobaeza) [inline website_portal_purchase/__openerp__.py]: s/invoices/supplier invoices

File: website_portal_purchase/README.rst [hunk 8]
```diff
@@ -0,0 +1,64 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+
+============================
+Website Portal for Purchases
+============================
+
+This module add the user's purchases documents in the frontend portal.
+Suppliers will be able to see the list and state of their quotations, purchases
```
Comment (pedrobaeza) [inline website_portal_purchase/README.rst]: s/quotations/request for quotations

Comment (pedrobaeza) [inline website_portal_purchase/README.rst]: s/purchase/purchase orders

File: website_portal_purchase/controllers/main.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        invoiced_lines = request.env['account.invoice.line'].search([
+            ('invoice_id', 'in', order.invoice_ids.ids)
+        ])
+        order_invoice_lines = {il.product_id.id: il.invoice_id
+                               for il in invoiced_lines}
+
+        return request.website.render(
+            "website_portal_purchase.orders_followup",
+            {
+                'order': order.sudo(),
```
Comment (pedrobaeza) [inline website_portal_purchase/controllers/main.py]: Why sudo here? This is too dangerous. You should adjust security for portal users, but not allow to access any linked record

File: website_portal_purchase/controllers/main.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            'date': datetime.date.today().strftime('%Y-%m-%d'),
+            'quotations': quotations,
+            'orders': orders,
+            'invoices': invoices,
+        })
+        return response
+
+    @http.route(['/my/supplier/orders/<int:order>'], type='http', auth="user",
+                website=True)
+    def supplier_orders_followup(self, order=None):
```
Comment (pedrobaeza) [inline website_portal_purchase/controllers/main.py]: call it `order_id`

File: website_portal_purchase/views/website_portal_purchase.xml [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+                    <th>Purchase Order #</th>
+                    <th>Confirmation Date</th>
+                    <th></th>
+                    <th>Total</th>
+                  </tr>
+                </thead>
+                <t t-foreach="orders" t-as="order">
+                    <tr t-att-class="'hidden to_hide' if order.state not in ['manual', 'progress'] and order_index &gt; 4 else ''">
+                        <td>
+                            <a t-att-href="'/my/supplier/orders/'+str(order.id)"><t t-esc="order.name"/></a>
```
Comment (pedrobaeza) [inline website_portal_purchase/views/website_portal_purchase.xml]: You can include here also a link to download the purchase order report

File: website_portal_purchase/views/website_portal_purchase.xml [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+        <div id="wrap">
+            <div class="container">
+              <div class="oe_structure">
+                <div class="row">
+                    <div class="col-sm-4">
+                        <ol class="breadcrumb mt8">
+                            <li><a href="/my/supplier/home">My Account</a></li>
+                            <li>
+                              <t t-if="order.state == 'sent'">
+                                  Quotation
```
Comment (pedrobaeza) [inline website_portal_purchase/views/website_portal_purchase.xml]: Request for Quotation

File: website_portal_purchase/views/website_portal_purchase.xml [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+      <t t-call="website.layout">
+        <div id="wrap">
+            <div class="container">
+              <div class="oe_structure">
+                <div class="row">
+                    <div class="col-sm-4">
+                        <ol class="breadcrumb mt8">
+                            <li><a href="/my/supplier/home">My Account</a></li>
+                            <li>
+                              <t t-if="order.state == 'sent'">
```
Comment (pedrobaeza) [inline website_portal_purchase/views/website_portal_purchase.xml]: Include also state='draft'

File: website_portal_purchase/views/website_portal_purchase.xml [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+              </div>
+              <div id="optional_placeholder"></div>
+              <div class="container">
+
+                  <div class="panel panel-default">
+                    <div class="panel-heading">
+                      <div class="row">
+                        <div class="col-md-12">
+                          <h4>
+                            <t t-if="order.state == 'sent'">
```
Comment (pedrobaeza) [inline website_portal_purchase/views/website_portal_purchase.xml]: Also draft

File: website_portal_purchase/views/website_portal_purchase.xml [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+                              Quotation
```
Comment (pedrobaeza) [inline website_portal_purchase/views/website_portal_purchase.xml]: Request for Quotation

**Comment by rafaelbn** [comment]
ping @sergio-incaser

File: website_portal_purchase/views/website_portal_purchase.xml (hunk 13 above)
Comment (sergio-teruel) [inline website_portal_purchase/views/website_portal_purchase.xml]: I think that until not send it to supplier.

File: website_portal_purchase/views/website_portal_purchase.xml (hunk 14 above)
Comment (sergio-teruel) [inline website_portal_purchase/views/website_portal_purchase.xml]: I think that until not send it to supplier.

Comment (pedrobaeza) [inline website_portal_purchase/views/website_portal_purchase.xml]: One purchase order is a quotation in 'draft' and 'sent' states

**Comment by sergio-teruel** [comment]
@pedrobaeza Chages made.
@rafaelbn Ready for new review

**Comment by rafaelbn** [comment]
Hi,
- I have registered as "customer" and named the user "Is Customer"
- I have registered as "supplier" and name the user "Is Supplier"

As you can see, both can see "My account" and "My supplier account".
1. The first I don't understand is why there is a "My account" (website_portal_sale) if the module is not available? (it is still a PR: https://github.com/OCA/website/pull/111)
2. [FIX] I don't think Customer must see "My supplier account", in fact if you try to access you get an error "Forbidden".
3. [FIX] I don't think Supplier must see "My account"

What do you think?

**Comment by sergio-teruel** [comment]
@rafaelbn Changes done.

**1.-** That is not correct. 'My account' is added by _'website_portal'_ and the other modules adds his options.
**2.-** In this module we can show 'My Supplier Account' only if user is a supplier. --> **OK**
**3.-** That change would have to do it in the 'website_portal' module. For me ok.

**Comment by pedrobaeza** [comment]
I think my account is intended for a general purpose container for all partners (customer, suppliers, or none of both checked), so maybe you shouldn't create a new menu, but put the corresponding lists inside the "My Account" menu. The only doubt is the sales order and sales invoices sections are hidden when the partner is not a customer. If not, maybe this is the change you should propose to website_portal.

**Comment by sergio-teruel** [comment]
Only one link to My Account
captura de pantalla de 2015-11-18 11 29 38

File: website_portal_purchase/views/website_portal.xml [hunk 16]
```diff
@@ -0,0 +1,47 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+<data>
+
+    <template id="account" name="Account" inherit_id="website_portal.account">
+        <xpath expr="//ul[contains(@name,'account-content-nav')]" position="inside">
```
Comment (yajo) [inline website_portal_purchase/views/website_portal.xml:6]: This line seems to fail in runbot. Can you fix it please?

**Comment by sergio-teruel** [comment]
@Yajo rebase done.

**Comment by yajo** [comment]
54 files changed? O_o

I think you rebased on 9.0 or something like that. There are lots of translations for unrelated modules. Can you double-check, please?

**Comment by sergio-teruel** [comment]
@Yajo Rebase Fixed

**Comment by yajo** [comment]
Seems like you lost https://github.com/OCA/website/pull/90#issuecomment-153032787 with the rebase. Also please add tests, this adds lots of functionality that needs them.

**Comment by rafaelbn** [comment]
Travis is failing

**Comment by yajo** [comment]
Yes, that's why.

**Comment by rafaelbn** [comment]
Please @sergio-incaser try to review this PR, we must try to review againg, fix runbot and travis, thanks

**Comment by rafaelbn** [comment]
Could we fix travis to get this PR merged?

**Comment by rafaelbn** [comment]
ping @sergio-incaser @carlos-incaser @Yajo

**Comment by rafaelbn** [comment]
This PR I think is ready to be merge, just need travis fixed

Please @sergio-incaser @carlos-incaser @Yajo @pedrobaeza @antespi last review before merging

Functionally tested :+1:

**Comment by antespi** [comment]
Please add an oca_dependencies.txt file for fixing Travis, like this one : https://github.com/OCA/sale-workflow/blob/8.0/oca_dependencies.txt (only server-tools line)

**Comment by carlosdauden** [comment]
Add oca_dependencies to travis errors fixed

File: website_portal_purchase/views/website.xml [hunk 17]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+<data>
+
+    <template id="assets_frontend" inherit_id="website.assets_frontend" name="Website Portal Sale">
+      <xpath expr="." position="inside">
+        <link rel='stylesheet' href='/website_portal_purchase/static/src/less/website_portal_purchase.less'/>
```
Comment (yajo) [inline website_portal_purchase/views/website.xml]: I'd say this is not less bus just plain CSS. Please rename the file extension and this link.

**Comment by yajo** [comment]
Please provide Spanish translation.

**Comment by yajo** [comment]
Runbot's :x: are https://github.com/OCA/server-tools/issues/314 and https://github.com/odoo/odoo/issues/10159, so don't worry about it for now.

**Comment by sergio-teruel** [comment]
@Yajo changes done

File: website_portal_purchase/README.rst [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+Usage
+=====
+
+To use this module, you need to:
+
+1. Assign auth user group to the user.
+2. If you install 'portal_sale' you can invite user to access it.
+3. User can access to his request quotations, orders and supplier invoices
+from his account in user account link.
```
Comment (yajo) [inline website_portal_purchase/README.rst]: In .rst files, these lines must be aligned with their parent, so you need to put 2 paces before this line.

**Comment by yajo** [comment]
Reviewed, you are 2 spaces away from my thumb :wink:

**Comment by sergio-teruel** [comment]
@Yajo I add two spaces. Changes done..

**Comment by yajo** [comment]
Well seems they were 3, and you removed it from the lines that were right. [It still does not render properly](https://github.com/incaser/website/blob/43db5dbcc26d162a62e75674c93e784e0f88ff9d/website_portal_purchase/README.rst#usage). Please review [the docs](http://docutils.sourceforge.net/docs/user/rst/quickref.html#enumerated-lists).

**Comment by sergio-teruel** [comment]
@Yajo  sorry ... :blush:

**Comment by yajo** [comment]
:+1:

**Pull request merged**